nucypher/blockchain/eth/sol/source/contracts/MiningAdjudicator.sol

```diff
@@ -0,0 +1,485 @@
+    uint256 public percentagePenaltyCoefficient;
+    uint256 public rewardCoefficient;
+    mapping (address => uint256) public penaltyHistory;
+    mapping (bytes32 => bool) public evaluatedCFrags;
+
+    /**
+    * @param _escrow Escrow contract
+    * @param _hashAlgorithm Hashing algorithm
+    * @param _basePenalty Base for the penalty calculation
+    * @param _penaltyHistoryCoefficient Coefficient for calculating the penalty depending on the history
+    * @param _percentagePenaltyCoefficient Coefficient for calculating the percentage penalty
+    * @param _rewardCoefficient Coefficient for calculating the reward
+    **/
+    constructor(
+        MinersEscrow _escrow,
+        SignatureVerifier.HashAlgorithm _hashAlgorithm,
+        uint256 _basePenalty,
+        uint256 _penaltyHistoryCoefficient,
+        uint256 _percentagePenaltyCoefficient,
+        uint256 _rewardCoefficient
+    )
+        public
+    {
+        require(address(_escrow) != address(0) &&
+            _percentagePenaltyCoefficient != 0 &&
+            _rewardCoefficient != 0);
+        escrow = _escrow;
+        hashAlgorithm = _hashAlgorithm;
+        basePenalty = _basePenalty;
+        percentagePenaltyCoefficient = _percentagePenaltyCoefficient;
+        penaltyHistoryCoefficient = _penaltyHistoryCoefficient;
+        rewardCoefficient = _rewardCoefficient;
+    }
+
+    /**
+    * @notice Submit proof that miner created wrong CFrag
+    * @param _capsuleBytes Serialized capsule
+    * @param _capsuleSignatureByRequester Signature of Capsule by requester
+    * @param _capsuleSignatureByRequesterAndMiner Signature of Capsule by requester and miner
+    * @param _cFragBytes Serialized CFrag
+    * @param _cFragSignatureByMiner Signature of CFrag by miner
+    * @param _requesterPublicKey Requester's public key that was used to sign Capsule
+    * @param _minerPublicKey Miner's public key that was used to sign Capsule and CFrag
+    * @param _minerPublicKeySignature Signature of public key by miner's eth-key
+    * @param _preComputedData Pre computed data for CFrag correctness verification
+    **/
+    // TODO add way to slash owner of UserEscrow contract
+    function evaluateCFrag(
+        bytes memory _capsuleBytes,
+        bytes memory _capsuleSignatureByRequester,
+        bytes memory _capsuleSignatureByRequesterAndMiner,
+        bytes memory _cFragBytes,
+        bytes memory _cFragSignatureByMiner,
+        bytes memory _requesterPublicKey,
+        bytes memory _minerPublicKey,
+        bytes memory _minerPublicKeySignature,
+        bytes memory _preComputedData
+    )
+        public
+    {
+        require(_minerPublicKey.length == 65 && _requesterPublicKey.length == 65);
+
+        // Check that CFrag is not evaluated yet
+        bytes32 evaluationHash = SignatureVerifier.hash(
+            abi.encodePacked(_capsuleBytes, _cFragBytes), hashAlgorithm);
+        require(!evaluatedCFrags[evaluationHash]);
```

Comment:
Forgive my ignorance here.  So, the `evaluatedCFrags` mapping will change any time a block includes a completed transaction against this contract, right?

So what happens if the same CFrags is challenged many times in the same block time, before the adjudication is finalized in a block?  Will some of the adjudications end up in Uncles or thrown away entirely?

Sub-question:  If an Uncle contains a proper adjudication but the block itself didn't include it, is there a way to use the adjudication result on the Uncle?

Reply:
First transaction in block will be completed and others will fail because of `require(!evaluatedCFrags[evaluationHash]);` (state changes after each transaction in a block). So others can't make any changes on state. The same as we put them in different blocks.

I dont know if miner could use precalculated result from uncle, but for client it's useless. Because uncle does not change anything for client, and result from this transaction just could show is CFrag correct. But client can check it locally without transaction.

Reply:
Yes, the client can check locally. There could be several competing transactions in the block though. If there are, the first will collect the reward, and others will pat small fee for checking this condition before fail.
That's not ideal, but probably ok for now